pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go

```diff
@@ -1450,6 +1492,137 @@ func TestGetPipelineConditionStatus(t *testing.T) {
+		Type:    apis.ConditionSucceeded,
+		Status:  corev1.ConditionFalse,
+		Reason:  v1beta1.PipelineRunReasonFailed.String(),
+		Message: getExpectedMessage(corev1.ConditionFalse, 1, 0, 0, 1, 0),
+	}
+	if d := cmp.Diff(wantCondition, c); d != "" {
+		t.Fatalf("Mismatch in condition for pipeline with one failed DAG task and failed final task %s", diff.PrintWantGot(d))
+	}
+}
+
+func TestGetPipelineConditionStatus_OneDAGFailedNoFinalStarted(t *testing.T) {
+
+	// pipeline state with one DAG failed, no final started
+	state := PipelineRunState{{
+		TaskRunName:  "task0taskrun",
+		PipelineTask: &pts[0],
+		TaskRun:      makeFailed(trs[0]),
+	}, {
+		TaskRunName:  "notRunningTaskRun",
+		PipelineTask: &pts[1],
+		TaskRun:      nil,
+	}}
+
+	d, err := dag.Build(v1beta1.PipelineTaskList([]v1beta1.PipelineTask{*state[0].PipelineTask}))
+	if err != nil {
+		t.Fatalf("Unexpected error while buildig DAG for state %v: %v", state, err)
+	}
+
+	df, err := dag.Build(v1beta1.PipelineTaskList([]v1beta1.PipelineTask{*state[1].PipelineTask}))
+	if err != nil {
+		t.Fatalf("Unexpected error while buildig DAG for state %v: %v", state, err)
+	}
+
+	pr := tb.PipelineRun("somepipelinerun")
+	c := GetPipelineConditionStatus(pr, state, zap.NewNop().Sugar(), d, df)
+	wantCondition := &apis.Condition{
+		Type:    apis.ConditionSucceeded,
+		Status:  corev1.ConditionUnknown,
+		Reason:  v1beta1.PipelineRunReasonRunning.String(),
+		Message: getExpectedMessage(corev1.ConditionUnknown, 0, 1, 0, 1, 0),
+	}
+	if d := cmp.Diff(wantCondition, c); d != "" {
+		t.Fatalf("Mismatch in condition for pipeline with one failed DAG task, no final started %s", diff.PrintWantGot(d))
+	}
+}
+
+func TestGetPipelineConditionStatus_OneDAGFailedFinalFailed(t *testing.T) {
+
+	// pipeline state with one DAG failed, one final task failed
+	state := PipelineRunState{{
+		TaskRunName:  "task0taskrun",
+		PipelineTask: &pts[0],
+		TaskRun:      makeFailed(trs[0]),
+	}, {
+		TaskRunName:  "notRunningTaskRun",
+		PipelineTask: &pts[1],
+		TaskRun:      makeFailed(trs[0]),
+	}}
+
+	d, err := dag.Build(v1beta1.PipelineTaskList([]v1beta1.PipelineTask{*state[0].PipelineTask}))
+	if err != nil {
+		t.Fatalf("Unexpected error while buildig DAG for state %v: %v", state, err)
+	}
+
+	df, err := dag.Build(v1beta1.PipelineTaskList([]v1beta1.PipelineTask{*state[1].PipelineTask}))
+	if err != nil {
+		t.Fatalf("Unexpected error while buildig DAG for state %v: %v", state, err)
+	}
+
+	pr := tb.PipelineRun("somepipelinerun")
+	c := GetPipelineConditionStatus(pr, state, zap.NewNop().Sugar(), d, df)
+	wantCondition := &apis.Condition{
+		Type:    apis.ConditionSucceeded,
+		Status:  corev1.ConditionFalse,
+		Reason:  v1beta1.PipelineRunReasonFailed.String(),
+		Message: getExpectedMessage(corev1.ConditionFalse, 0, 0, 0, 2, 0),
+	}
+	if d := cmp.Diff(wantCondition, c); d != "" {
+		t.Fatalf("Mismatch in condition for pipeline with one failed DAG task and failed final task %s", diff.PrintWantGot(d))
+	}
+}
```

Comment:
it feels like a lot of these functions could be combined into table driven tests, what do you think @pritidesai ? it would be easier to compare the test cases to each other

i think separate tests make sense when the test assert logic has to be different, but if everything about the setup, function call and assertions is the same expect the inputs and expected values, table driven tests really help a lot

Reply:
you are right @bobcatfish, it does make sense and makes it simpler, I have combined these tests into one table driven test.